Implement truncated normal with box-muller

mrTsjolder opened the pull request:
This is an implementation of the box-muller transform for a normal distribution truncated at two standard deviations (as requested in issue #6381).

I did not quite get why there would be need to distinguish between integer and tensor sizes, given that theano optimises the graph and probably includes constant folding. Therefore, I have introduced my own interpretation of the normal code (which I personally find more readable, if I'm allowed to be smug for a second). If one of two approaches turn out to be better/faster/more readable/more whatever, quite a bit of code can be reused and a refactoring might be advisable.

nouiz commented:
The reason why we generate a simpler graph for constant is that it make reading the not optimized graph easier and make the compilation faster. It isn't stricly requested, but is nice to do.

This would need test in the file theano/sandbox/tests/test_rng_mrg.py to verify is the generated distribution is well clipped.

nouiz commented on theano/sandbox/rng_mrg.py:
```diff
@@ -1117,6 +1118,79 @@ def normal(self, size, avg=0.0, std=1.0, ndim=None,
+        to_fix0 = nonzero(to_fix0)[0]
+        to_fix1 = nonzero(to_fix1)[0]
+        n_fix_samples = to_fix0.size + to_fix1.size
+        u_fix = self.uniform(size=(n_fix_samples, ), low=0.13585, high=1.,
```
From the wikipedia page, I think the "low" number would be dtype dependent.

I would be interrested to know how the other framework does it. Did you looked at that? I checked tensorflow and it a variable number of time until it accept a number. The way you do it seem interresting and could prevent that need.

I would like the input from @lamblin and @abergeron on the way you implement it. resample only once with with a different uniform to force a max of 2 stddev.

mrTsjolder replied:
In what sense would `low` depend on the `dtype`? If `low = 0.`, the `dtype` will indeed cause tail-truncation (that is where my idea for the resampling came from).

Which other framework? Keras simply samples from a normal and clips all values to the edges, leaving a distribution that is very non-Gaussian and severely messes with the variance. Tensorflow does indeed some iterative resampling and I have no idea where the CNTK implementation is hidden.

mrTsjolder commented:
I would argue that it is more valuable to have readable code than to have readable graphs.  If there are really problems with readability of optimised graphs, it should be solved where the origin of these problems lie. Note that this is just a personal opinion of mine.

As to the requested changes: What am I supposed to change? Is it just the test case or are other changes necessary (e.g. refactoring)? Strictly speaking, it is probably not completely correct, but I think it does well for the 5% of data that is "out of bounds".

mrTsjolder commented:
I noticed that there is a `numpy`-based random generator as well. `numpy` does not offer a truncated normal distribution (yet). I proposed a `numpy` implementation of the box-muller transform in #6381, which could be used. Another option, would be to use random number generators from `scipy.stats`, but I am not sure whether that is an option.

nouiz commented:
I don't think it is useful to implement it in the numpy based random generator. It is slower and don't work on GPU.

abergeron commented on theano/sandbox/rng_mrg.py:
```diff
@@ -1117,6 +1118,81 @@ def normal(self, size, avg=0.0, std=1.0, ndim=None,
         return final_samples
 
+    def truncated_normal(self, size, avg=0.0, std=1.0, ndim=None,
+                         dtype=None, nstreams=None):
```
This would need to add support for the new `target` parameter: https://github.com/Theano/Theano/pull/6355

mrTsjolder replied:
@nouiz This is the review of abergeron I referred to.

mrTsjolder commented:
In my enthusiasm, I changed quite a bit of code and added an (inefficient) implementation for binomial distributions for n > 1. I have also made it possible to get 0-dimensional sizes (which happens to be convenient for the normal distribution). I also tried to improve the documentation and overall consistency. I hope this is no issue

The tensorflow implementation has been made available through the `truncate`-flag in `normal`. `truncated_normal` returns samples from a truncated normal with the provided standard deviation.

As to the failing circleci-test: could it be that this is not due to code that I implemented?

abergeron commented:
You can ignore CircleCI for this PR, the base branch did not have the proper config.

notoraptor commented on theano/sandbox/rng_mrg.py:
```diff
@@ -1029,96 +1048,177 @@ def multinomial_wo_replacement(self, size=None, n=1, pvals=None,
+        --------
+        normal
+        """
+        std = std / tensor.constant(.87962566103423978)
```
Is there a mathematical explanation/expression for that number (`.87962566103423978`) ? If so, could you use it instead of putting directly the number value ? As you do in line 1118 with `lower = tensor.constant(1. / np.e**2, dtype=dtype)`.

mrTsjolder replied:
It is the result of `scipy.stats.truncnorm.std(a=-2, b=2, loc=0., scale=1.)` (see comment in `test_normal_truncation`). I was not sure whether this is worth a scipy import or whether Theano already depends on scipy. Should I replace the constant by the above mentioned line?

lamblin replied:
Theano has no hard dependency on scipy, so it's better not to import it here.
If you could just copy the comment here as well, I think that should be enough.

notoraptor replied:
I agree with @lamblin , just adding a comment to tell how it is calculated should be sufficient.

notoraptor replied:
PS: Ok, comment is in `test_normal_truncation`. Could you add it here too, in `truncated_normal()` function ?

notoraptor commented on theano/sandbox/rng_mrg.py:
```diff
@@ -804,101 +803,121 @@ def pretty_return(self, node_rstate, new_rstate, sample, size, nstreams):
 
-        assert r.dtype == dtype
-        return r
+    def binomial(self, size, n=1, p=0.5,
```
There are currently no tests for `binomial()` when `n > 1`, in `theano/sandbox/tests/test_rng_mrg.py`. Could you add a test for that case ?

notoraptor commented on theano/sandbox/rng_mrg.py:
```diff
@@ -804,101 +803,121 @@ def pretty_return(self, node_rstate, new_rstate, sample, size, nstreams):
+            A Theano tensor of samples randomly drawn from a binomial distribution.
+
+        """
+        if isinstance(n, (np.integer, integer_types)):
```
What if `n` is a floating python or numpy value ?

notoraptor commented on theano/sandbox/rng_mrg.py:
```diff
@@ -887,11 +887,12 @@ def binomial(self, size, n=1, p=0.5,
         """
-        if isinstance(n, (np.integer, integer_types)):
+        # TODO: use ABCs for these checks?
+        if not isinstance(p, theano.Variable):
```
I think it should be `n` instead of `p` here, isn't it ?

Also, maybe add after that line something like:
```python
assert n == int(n), "n must be an integer"
```

notoraptor commented on theano/sandbox/tests/test_rng_mrg.py:
```diff
@@ -321,60 +321,62 @@ def test_broadcastable():
-    f = theano.function(var_input, u)
-    f(*input)
+    for size, const_size, var_input, input, n, p, rtol, std_tol in test_cases:
+        print(const_size, n, p, rtol, std_tol)
```
This print should be removed for final code.

notoraptor commented on theano/sandbox/rng_mrg.py:
```diff
@@ -887,15 +887,18 @@ def binomial(self, size, n=1, p=0.5,
-            raise ValueError("Probability of success must be between 0 and 1 for a binomial distribution", p)
+        if not isinstance(p, theano.Variable):
+            assert p == float(p), "p must be a float"
+            if not 0. < p < 1.:
```
p can be 0 or 1, so it should be `if not 0. <= p <= 1.` instead !

mrTsjolder replied:
I would like to apologise for my idiocy...

nouiz commented on theano/sandbox/rng_mrg.py:
```diff
@@ -804,101 +803,125 @@ def pretty_return(self, node_rstate, new_rstate, sample, size, nstreams):
-                nstreams=None, **kwargs):
-        # TODO : need description for parameter 'size', 'ndim', 'nstreams'
+    def uniform(self, size, low=0.0, high=1.0,
+                ndim=None, dtype=None, nstreams=None, target=None):
```
Why the change about target? Now it is different then the other method and I would prefer to not change this. So I would like to know why you changed it.

mrTsjolder replied:
Abergon requested support for the `target` parameter.

nouiz replied:
ok, but he was referencing to adding its support to truncated_normal, not change the interface here. Before you changed this code, it was an example of how to add that support.

mrTsjolder replied:
@nouiz In that case: this used to be a `**kwargs` that could only get the keyword argument `'target'` and was not passed to any other function. I considered that, in this case, it would be cleaner to make `target` a direct keyword argument.

Should I revert this change?

nouiz replied:
It is better to have all method implemented the same. Also, the fact of passing `target=None` or not passing target don't do exactly the same thing. This is why it was done like this. We should not change the behavior here.

nouiz commented on theano/sandbox/rng_mrg.py:
```diff
@@ -804,101 +803,125 @@ def pretty_return(self, node_rstate, new_rstate, sample, size, nstreams):
 
-        low = cast(low, dtype=dtype)
-        high = cast(high, dtype=dtype)
+        low = tensor.cast(low, dtype=dtype)
```
This change the graph. Now the undefined_grad is swapped comparted to cast in the graph. I don't think this is a problem, but I prefer to raise it in case someone think it is a problem.

mrTsjolder replied:
This change was introduced to improve the consistency of the code. I used the original implementation of `MRG_RandomStreams.normal` as a reference. Just to provide some more background on where this comes from.

nouiz replied:
So this confirm it should not be a problem. thanks.

nouiz commented on theano/sandbox/rng_mrg.py:
```diff
@@ -13,17 +13,15 @@
+from theano.sandbox import multinomial
+
+floatX = theano.config.floatX
+intX = floatX.replace('float', 'int')
```
I don't like this change. This change the user interface just before a release and isn't good. If you want this PR to be included in the next release, can you keep in it just the changes about the new random generator?

Also, did you see the announce: https://groups.google.com/d/msg/theano-users/7Poq8BZutbY/rNCIfvAEAwAJ

This mostly tell we won't change the interface, so moving this interface change to another PR isn't useful.

mrTsjolder replied:
Would it be okay to use `_floatX` and `_intX` instead to make them *private*. That would leave the interface unchanged (assuming that private variables are not part of the interface).

Note: as you might notice, I am not a full-blood python developer...

nouiz replied:
This isn't the interface change that cause problem. The problem is that when you set floatX=float32, then the dtype change compared to before. In the past, but default it was always int64 and didn't looked at floatX. It should stay like that.

mrTsjolder replied:
@nouiz I see, but wouldn't it make more sense to have `int32` as default when `floatX` is `float32` and `int64` when `floatX` is `float64`? E.g. for people working on 32-bit hardware...

nouiz replied:
float32 isn't a questions of 32 bit hardware or not. You can want to use float32, but need int64 for indexing. There was a discussion to make an intX in the past. I forgot why we didn't do it, probably not enough time to get to it. But now isn't a good time to implement that.

nouiz commented on theano/sandbox/rng_mrg.py:
```diff
@@ -940,16 +969,16 @@ def multinomial(self, size=None, n=1, pvals=None, ndim=None, dtype='int64',
             raise NotImplementedError(("MRG_RandomStreams.multinomial only"
                                        " implemented for pvals.ndim = 2"))
 
-    def choice(self, size=1, a=None, replace=True, p=None, ndim=None,
```
You changed many thing unrelated to your PR. Can you remove those commits instead of reverting all the changes? Hre, nstreams must stay and the order of parameter must not change. Those 2 changes change the user interface.

mrTsjolder replied:
@nouiz I'm sorry

Does this mean I should also delete the binomial implementation for `n>1`?

`nstreams` is still there in some sense. It will just be passed through `kwargs`. I personally find it cleaner to use `kwargs` to simply pass keyword arguments to other functions. It also allows for easier extensibility (just add another `kwarg` to uniform and it will work). 

This being said, I will just do as you ask.

nouiz commented:
I removed the mostly duplicated test. We can merge when CI passes

nouiz commented:
jenkins test this

nouiz commented:
I ran the CPU random tests with the gpu enabled. This discovered one corner case in random (0 sized output) problem. I included the fix in this PR.

jenkins test this.

abergeron commented:
BTW you can ignore CircleCI failures for this PR.

nouiz commented:
jenkins test this

mrTsjolder commented:
@nouiz The test you deleted was not completely a duplicate. It was mainly there to assure that the standard deviation of the truncated normal is actually the required standard deviation, whereas the normal-truncation-test rather tests that the normal is truncated. I must admit that both tests are a little bit overkill.

It's not that serious, but I considered it would be better to avoid surprises.

nouiz commented:
jenkins test this

nouiz commented:
I did small modification and re-added  the test.

jenkins test this.

I think it can be merged when CI pass. thanks.

nouiz commented:
jenkins test this
